### DISCUSSION TITLE
Issues after upgrading 9 servers to latest version - "xxxx has incorrect configuration. Expected same MINIO_ environment variables and values"

### DISCUSSION THREAD

**Discussion by melsom** [minio/minio#13297] [Q&A] [answered]
Server minio1 reports:
Sep 24 13:46:45 minio1 minio[1008787]: API: SYSTEM()
Sep 24 13:46:45 minio1 minio[1008787]: Time: 13:46:45 CEST 09/24/2021
Sep 24 13:46:45 minio1 minio[1008787]: Error: https://minio2.XXXXXX.XX:9000/data1 as has incorrect configuration: Expected same MINIO_ environment variables and values (*fmt.wrapError)
Sep 24 13:46:45 minio1 minio[1008787]:        1: cmd/server-main.go:515:cmd.serverMain()
Server minio2 starts normal, reporting all buckets and drives online, but after a minute or two:
Sep 24 12:54:20 minio2 minio[1031899]: API: SYSTEM()
Sep 24 12:54:20 minio2 minio[1031899]: Time: 12:54:20 CEST 09/24/2021
Sep 24 12:54:20 minio2 minio[1031899]: DeploymentID: bbd7798b-f730-401a-b5eb-38e7a3a5a980
Sep 24 12:54:20 minio2 minio[1031899]: Error: Server not initialized, please try again (*errors.errorString)
Sep 24 12:54:20 minio2 minio[1031899]:        1: cmd/erasure.go:499:cmd.erasureObjects.nsScanner.func3()
No config has been changed. I even diffed them and the only thing that is different is "minio1.%hostname%" vs "minio2.%hostname%" in the "MINIO_OPTS" variable:

Minio1
MINIO_OPTS="--address minio1.XXXX.XX:9000 --console-address minio1.XXXX.XX:9001 https://minio{1...3}.XXXX.XX/data{1...12} https://minio{4...6}.XXXX.XX/data{1...12} https://minio{7...9}.XXXX.XX/data{1...12}"

Minio2:
MINIO_OPTS="--address minio2.XXXX.XX:9000 --console-address minio2.XXXX.XX:9001 https://minio{1...3}.XXXX.XX/data{1...12} https://minio{4...6}.XXXX.XX/data{1...12} https://minio{7...9}.XXXX.XX/data{1...12}"

Other variables are identical except "MINIO_SERVER_URL" which specifies each servers console URL, https://minio%num%.%domain%:9000

Running directly on servers, not in docker. Other servers, minio3 to 9 reports minio2 drives as online. Each server has 12 drives. Minio1 seems to only be reporting this error on /data1.

Any clue as to why this is happening?

**Comment by vadmeste** [comment, marked as answer]
@melsom you can avoid putting server address for --address and --console-address, that way they will be identical.

For MINIO_SERVER_URL, you can use the load balancer address, so it will be identical as well.

**Reply by melsom** [reply]
Okay, so you suggest OPTS to be:
--address :9000 --console-address :9001 https://minio{1...3}.XXXX.XX/data{1...12} https://minio{4...6}.XXXX.XX/data{1...12} https://minio{7...9}.XXXX.XX/data{1...12}

If I understand you correctly? 

All 9 servers have this setup, and I see from the code that it will halt on first diff of opts. Is this something new in this minio version? Just thinking it would be a breaking change for others aswell.

**Reply by melsom** [reply]
Looks like this is new in the code yes, but shouldn't also MINIO_SERVER_URL be excluded? It doesn't have to be the same across all nodes.

**Reply by melsom** [reply]
It even triggers on MINIO_CACHE_DRIVES. Shouldn't the node be able to have different number of cache drives?

**Reply by melsom** [reply]
@vadmeste Seems I am in a bit of a squeeze now. 6 of 9 systems have 2 cache drives. The last 3 have only 1, due to size difference. Have worked fine up until now. Trying to either remove or add cache results in an error.

Removing a cache drive to even the number:
Sep 24 15:00:47 minio6 minio[231540]: Waiting for all MinIO sub-systems to be initialized.. lock acquired
Sep 24 15:00:47 minio6 minio[231540]: Verifying if 23 buckets are consistent across drives...
Sep 24 15:00:47 minio6 minio[231540]: Automatically configured API requests per node based on available memory on the system: 313
Sep 24 15:00:47 minio6 minio[231540]: All MinIO sub-systems initialized successfully
Sep 24 15:00:47 minio6 minio[231540]: Waiting for all MinIO IAM sub-system to be initialized.. lock acquired
Sep 24 15:00:48 minio6 minio[231540]: API: SYSTEM()
Sep 24 15:00:48 minio6 minio[231540]: Time: 15:00:48 CEST 09/24/2021
Sep 24 15:00:48 minio6 minio[231540]: DeploymentID: bbd7798b-f730-401a-b5eb-38e7a3a5a980
Sep 24 15:00:48 minio6 minio[231540]: Error: Expected number of cache drives 2 , got  1 (*errors.errorString)
Sep 24 15:00:48 minio6 minio[231540]:        5: cmd/format-disk-cache.go:304:cmd.validateCacheFormats()
Sep 24 15:00:48 minio6 minio[231540]:        4: cmd/format-disk-cache.go:341:cmd.loadAndValidateCacheFormat()
Sep 24 15:00:48 minio6 minio[231540]:        3: cmd/disk-cache.go:570:cmd.newCache()
Sep 24 15:00:48 minio6 minio[231540]:        2: cmd/disk-cache.go:757:cmd.newServerCacheObjects()
Sep 24 15:00:48 minio6 minio[231540]:        1: cmd/server-main.go:571:cmd.serverMain()
Sep 24 15:00:48 minio6 minio[231540]: ERROR Unable to initialize disk caching: Expected number of cache drives 2 , got  1


Adding a cache drive to even the number:
ep 24 14:57:13 minio7 minio[237909]: Waiting for all MinIO sub-systems to be initialized.. lock acquired
Sep 24 14:57:13 minio7 minio[237909]: Verifying if 23 buckets are consistent across drives...
Sep 24 14:57:13 minio7 minio[237909]: Automatically configured API requests per node based on available memory on the system: 306
Sep 24 14:57:13 minio7 minio[237909]: All MinIO sub-systems initialized successfully
Sep 24 14:57:13 minio7 minio[237909]: Waiting for all MinIO IAM sub-system to be initialized.. lock acquired
Sep 24 14:57:13 minio7 minio[237909]: API: SYSTEM()
Sep 24 14:57:13 minio7 minio[237909]: Time: 14:57:13 CEST 09/24/2021
Sep 24 14:57:13 minio7 minio[237909]: DeploymentID: bbd7798b-f730-401a-b5eb-38e7a3a5a980
Sep 24 14:57:13 minio7 minio[237909]: Error: Expected number of cache drives 1 , got  2 (*errors.errorString)
Sep 24 14:57:13 minio7 minio[237909]:        5: cmd/format-disk-cache.go:304:cmd.validateCacheFormats()
Sep 24 14:57:13 minio7 minio[237909]:        4: cmd/format-disk-cache.go:341:cmd.loadAndValidateCacheFormat()
Sep 24 14:57:13 minio7 minio[237909]:        3: cmd/disk-cache.go:570:cmd.newCache()
Sep 24 14:57:13 minio7 minio[237909]:        2: cmd/disk-cache.go:757:cmd.newServerCacheObjects()
Sep 24 14:57:13 minio7 minio[237909]:        1: cmd/server-main.go:571:cmd.serverMain()
Sep 24 14:57:13 minio7 minio[237909]: ERROR Unable to initialize disk caching: Expected number of cache drives 1 , got  2
Sep 24 14:57:13 minio7 systemd[1]: minio.service: Main process exited, code=exited, status=1/FAILURE
Sep 24 14:57:13 minio7 systemd[1]: minio.service: Failed with result 'exit-code'.
Sep 24 14:57:14 minio7 systemd[1]: minio.service: Scheduled restart job, restart counter is at 1.

**Reply by melsom** [reply]
Sorry for all these replies. I have temporarily resolved it by recompiling minio and adding MINIO_CACHE_DRIVES and MINIO_SERVER_URL to cmd/bootstrap-peer-server.go. The last one isn't strictly necessary but works better like that for my setup. Its something that should be added to the repo I believe unless there is something I've misunderstood! :)

**Reply by vadmeste** [reply]
> Sorry for all these replies. I have temporarily resolved it by recompiling minio and adding MINIO_CACHE_DRIVES and MINIO_SERVER_URL to cmd/bootstrap-peer-server.go. The last one isn't strictly necessary but works better like that for my setup. Its something that should be added to the repo I believe unless there is something I've misunderstood! :)

It is just better to have all of them to be the same in a distributed setup. It is easier, less prone to errors and more suitable to automation and Kubernetes.

**Reply by poornas** [reply]
You really don't need disk caching in front of a MinIO erasure/distributed setup here. Are the cache drives ssd's?

**Reply by melsom** [reply]
@poornas The drives in the nodes are HDD, the cache drives are NVMe storage. Since it would cache often accessed files it would take some of the load of the HDDs? 

If not, how can we remove the cache without MinIO complaining?

**Reply by poornas** [reply]
You can just unset MINIO_CACHE_DRIVES and restart MinIO - or if you still want to use the cache drives, clear your cache drives including all hidden files in .minio.sys and specify cache drives that match across the servers.

**Reply by melsom** [reply]
Okay thank you. Good to know. Would you consider adding MINIO_CACHE_DRIVES to the skipEnvs variable in cmd/bootstrap-peer-server.go for us to be able to continue have this uneven cache? We will eventually fix it but we will really like to keep on updating with an official release instead of compiling our own.

**Reply by harshavardhana** [reply]
You should avoid using MINIO_CACHE_DRIVES for distributed setup, it is not designed to be used with the distributed setup. Please only use it with Gateway.

**Reply by melsom** [reply]
Duly noted thanks :)
